### ISSUE TITLE
Infer_shape_partial for rank 0 arrays

### ISSUE THREAD

**Issue by aseyboldt** [apache/incubator-mxnet#8133]
In the python interface of mxnet it seems to be impossible to distinguish between an array of unknown shape and an array with known shape of `()`:
```python
a = mx.sym.var('a')
b = mx.sym.var('b')

(a + b).infer_shape_partial(b=())
([(), ()], [()], [])
```
In this case `b` is known to be a scalar, while `a`'s shape is unknown.
Shouldn't it return something like `([None, ()], [None], [])`?

If `a` is set to be a scalar, it returns an invalid result:
```python
a = mx.sym.var('a')
b = mx.sym.var('b')

(a + b).infer_shape_partial(a=(), b=(1, 2))
([(1, 2), (1, 2)], [(1, 2)], [])
```
It identifies the shape of `a` as `(1, 2)`, even though we set it to a scalar.

**Comment by szha** [comment]
I don't think we have the concept of scalar in ndarray or symbol. Shape of `()` doesn't mean it's a scalar, it means its shape is unknown and needs to be inferred.

**Comment by aseyboldt** [comment]
`ndarray` does allow scalars:
```python
>>> a = mx.nd.ones(())
>>> a.shape
()
>>> a.size
1
>>> a.asnumpy()
array(1.0, dtype=float32)
>>> mx.nd.ones((1,)).asnumpy()
array([ 1.], dtype=float32)
```
This seems to lead to the invalid memory access in the example above (which by the way I think is a serious bug).

I feel bad about criticising someone else's project after using it for only a couple of days, but I have to admit it is pretty much beyond me why you would design a tensor library without the concept of a scalar. :sweat_smile:
